### ISSUE TITLE
Add  specialized serializer for Scala

### ISSUE THREAD

**Issue by chaokunyang** [alipay/fury#682]

**Comment by pjfanning** [comment]
I would recommend that specialized Scala support be put in a separate jar from the Java support. The Scala Library jar is big so you don't want to have a dependency on it unless you need to. The fury-scala jar would need to be published for multiple Scala versions. Many Spark and Flink users still use Scala 2.11 so you would need to support 2.11, 2.12, 2.13 and 3. If you are not familiar with Scala, publishing the Scala version specific jars is pretty straightforward because build tools like Maven have specialized plugins for Scala.

**Comment by chaokunyang** [comment]
Maybe we can use `java.lang.invoke.MethodHandle` to load scala classes dynamically to avoid source dependency on scala library. In this way, we can avoid release different jar for different scala version too.

**Comment by pjfanning** [comment]
The problem is the scala-library jar dependency. You need a different scala-library jar depending on which Scala version is being used.

**Comment by chaokunyang** [comment]
Perhaps we can load it using reflection without introducing scala dependence? If so, we will release only one jar for all scala versions, which simply user dependence management too.

**Comment by pjfanning** [comment]
Supporting Serialization/Deserialization of Collection<T> is very hard with Java Reflection only. The T is erased when the T is a primitive type. Scala has its own separate reflection libraries but this is also messy because Scala 3 abandoned the scala-reflect lib available in Scala 2. New Scala 3 libs are being developed instead.

This FAQ shows some of the problems that Jackson-Module-Scala has in this area.

https://github.com/FasterXML/jackson-module-scala/wiki/FAQ

**Comment by chaokunyang** [comment]
> Supporting Serialization/Deserialization of Collection is very hard with Java Reflection only. The T is erased when the T is a primitive type. Scala has its own separate reflection libraries but this is also messy because Scala 3 abandoned the scala-reflect lib available in Scala 2. New Scala 3 libs are being developed instead.
> 
> This FAQ shows some of the problems that Jackson-Module-Scala has in this area.
> 
> https://github.com/FasterXML/jackson-module-scala/wiki/FAQ

I see, thanks. I didn't expect scala will erase primitive type for nested generic type:
![image](https://github.com/alipay/fury/assets/12445254/5b656a1c-c0ac-4b39-bd7e-ac6dfd4eda3c)

This breaks our assumption for java genercis. For java collection types, we can know the value is a `Long`, and write by long for it. 
```java
// generate pseudo-code
for (Entry e : map.entrySet()) {
  // null flag write
  stringSerailizer.writeString(e.getKey());
  buffer.writeSliLong(e.getValue());
} 
```

But for scala, since we can't know the actual type using relection when infer `Foo`  fiele generic type, we must write map value type foro every value, then write the the value:
```java
// generate pseudo-code
for (Entry e : map.entrySet()) {
  // null flag write
  stringSerailizer.writeString(e.getKey());
  Object v = e.getValue();
  writeType(v.getClass());
  Serializer s = getSerializer(v); // query from map
  s.write(buffer, v)// vritual method call
} 
```

It will be slower nad introduce more space overhead

**Comment by pjfanning** [comment]
It may only be necessary to write the inner type info for Collection<T> when the T is a primitive (long, int, etc.).

It may be possible to use scala libs to find to out what the erased types are but it is pretty complicated and you would like need a separate solution for Scala 2.x and Scala 3.

**Comment by pjfanning** [comment]
You could experiment with `List[java.lang.Long]`. You should find that the `java.lang.Long` is not erased. The erasure issue occurs for `List[scala.Long]` though. `scala.Long` is effectively an alias for the Java primitive type `long`.

Unfortunately, `scala.Long` is commonly used - so it is difficult to ignore the effect of this erasure.

**Comment by chaokunyang** [comment]
> It may only be necessary to write the inner type info for `Collection<T>` when the T is a primitive (long, int, etc.).
> 
> It may be possible to use scala libs to find to out what the erased types are but it is pretty complicated and you would like need a separate solution for Scala 2.x and Scala 3.

Our current protocol will write it only once, see #923 , but without this information ahead, the serialization for elements will introduce virtual method call, which will be slower. And the jit optimization is how fury get such boost up, it will be better if we can support such things for scala, considering it's used wisedly in spark/flink/akka.

**Comment by chaokunyang** [comment]
> You could experiment with `List[java.lang.Long]`. You should find that the `java.lang.Long` is not erased. The erasure issue occurs for `List[scala.Long]` though. `scala.Long` is effectively an alias for the Java primitive type `long`.
> 
> Unfortunately, `scala.Long` is commonly used - so it is difficult to ignore the effect of this erasure.

Yes, we shouldn't ignore this type information, otherwise the pperformance won't be the best

**Comment by chaokunyang** [comment]
Another thing I found it that scala collection doesn't implement java collection interface `Iterable/Collection/Set/Map`, which make the integration with fury collection JIT not easy to implement, since we don't have a java `Collection` base interface type to call to scala collection.

One method I can see is we convert scala collection to java collection in generated serializer(come with object creation overhead), or implement seperate jit support for such collection

**Comment by pjfanning** [comment]
Converting Scala collections to Java collections and vice versa will not be cheap.

**Comment by chaokunyang** [comment]
Now ith your new input, I totally agree  that we should add scala optimization support in a new library. This is a complicated work, and scala collections are much more complex than java collection framework. We must write the implementation using scala collection API, otherwise this work will be too much.

**Comment by chaokunyang** [comment]
Currently if a scala class doesn't have scala collections field type, the performance is good in fury. `case` type is supported by fury natively. But if collections types are included, the serialization performance will be not good since we can't use any collection generic information.  

Better scala serialization support should be done in a new jar. And perhaps we also need to add `a new adapter inferface` in java to make let scala collections hook into fury java codegen. Currently we do java codegen in `io.fury.builder.BaseObjectCodecBuilder#serializeForCollection`. Our codegen generate java code only. Maybe we need to adjust the jit and let it invoke the `new adapter inferface` to read or populate scala collection. In such wa, we can avoid generate scala code in our jit framework.
